### ISSUE TITLE
Support approving/rejecting individual files

### ISSUE THREAD

**Issue by inglesp** [opensafely-core/airlock#137]
To ensure that every file has been reviewed by two output checkers, output checkers need to be able to record, against each file in a release request, whether they have approved or rejected the file for release.  Output checkers need to be able to change their mind, so an approved file may later be rejected (or vice versa), as long as the file has not yet been released.

#136 describes how this fits in to the full release request lifecycle.

Depends on #125 (although maybe the UI could be tackled independently).

**Comment by inglesp** [comment]
Implementation notes:

To implement this, we'll need a new kind of object: a file approval belongs to a file, and records who approved/rejected a file and when.  There should only be one file approval per (file, reviewer) pair, so if a reviewer changes their mind, we should update the existing file approval.  It should only be possible to approve/reject a file of a request in the `SUBMITTED` state.

In detail...

* Rename `Status` enum to `RequestStatus`
* Create new `FileApprovalStatus` enum with members: `APPROVED` and `REJECTED`
* Update `StatusField` so that it is a base class, with two subclasses `RequestStatusField` and `FileApprovalStatusField`
* Create a `FileApproval` model in `local_db/models.py`
    * it needs the following fields:
        * `file`: a foreign key to `RequestFileMetadata`
        * `reviewer`: a text field with the reviewer's username
        * `status`: a `FileApprovalStatusField`
        * `created_at` and `updated_at`: see examples elsewhere
    * and a `unique_together` constraint on `file` and `reviewer`
* Add `approve_file` and `reject_file` methods to `LocalDBDataAccessLayer`
    * these should take a file and a user
    * they should either create a `FileApproval` for the given file and user, or update the existing one
* Add `approve_file` and `reject_file` methods to `BusinessLogicLayer` class
    * these should also take a file and a user, as well as a release request
    * they should check that the user is an output checker, and that the request's state is `SUBMITTED`
    * if the checks pass, they should call through to the corresponding methods on `LocalDBDataAccessLayer`
* Add `file_approve` and `file_reject` functions to `airlock/views/request.py`
    * these should only accept HTTP POSTs
    * these should call through to `approve_file` and `reject_file` in the business logic layer
* Add URLs for the new view functions
* Add "Approve" and "Reject" buttons to the release viewer page, but show only to output checkers and only when the request's state is `SUBMITTED`
    * If a user has already approved/rejected a file, only the button to reject/approve should be shown
* Add `get_file_approvals` method to `LocalDBDataAccessLayer`
    * this should take a file and return a list of dicts with the properties of file's approvals
* Add `FileApproval` dataclass to `business_logic.py` (cf `FileGroup` dataclass)
* Add `get_file_approvals` method to `BusinessLogicLayer` which calls through to the DAL and returns a `FileApproval`
* Support showing file approvals (who approved/rejected? when?) to the release viewer page

**Comment by rebkwok** [comment]
> Update StatusField so that it is a base class, with two subclasses RequestStatusField and FileApprovalStatusField
I don't think this is necessary, if `StatusField` allows choices to be specified when it's used in the model. `RequestMetadata` and `FileApproval` would both define a `status` field using `StatusField`, but with different choices.

> Add "Approve" and "Reject" buttons to the release viewer page, but show only to output checkers and only when the request's state is SUBMITTED
And don't show to output checkers if they are the author of the request

When a file is rejected, should we record the reason for the rejection (an additonal text field on the FileApproval model)? 

Is it a prerequisite for this ticket to make some decisions about how the logging will work? Do we need to show/track the history of the approval/rejection for a specific reviewer/file? (Maybe just simple logging for now is enough)

**Comment by inglesp** [comment]
Thanks @rebkwok, have updated the notes.

> When a file is rejected, should we record the reason for the rejection (an additonal text field on the FileApproval model)?

We'll want something like this, but don't need to implement that right now.  The next stepping stone will cover building out the workflow, including comments.

**Comment by bloodearnest** [comment]
> * Create new `FileApprovalStatus` enum with members: `APPROVED` and `REJECTED`

Can I suggest just `FileStatus`? I think the `Approval` is overly verbose.

> * Update `StatusField` so that it is a base class, with two subclasses `RequestStatusField` and `FileApprovalStatusField`
> * Create a `FileApproval` model in `local_db/models.py`

`FileReview`, or `FileVote` perhaps?  `FileApproval.status = rejected` doesn't quite scan for me?
     
>   * these should take a file and a user
>   * they should either create a `FileApproval` for the given file and user, or update the existing one
> * Add `approve_file` and `reject_file` methods to `BusinessLogicLayer` class
>   
>   * these should also take a file and a user, as well as a release request
>   * they should check that the user is an output checker, and that the request's state is `SUBMITTED`

They should also check that file is part of the request.

>   * if the checks pass, they should call through to the corresponding methods on `LocalDBDataAccessLayer`

call through to the `bll._dal` instance, which will be a LocalDBDataAccessLayer.

> * Add `file_approve` and `file_reject` functions to `airlock/views/request.py`
>   
>   * these should only accept HTTP POSTs
>   * these should call through to `approve_file` and `reject_file` in the business logic layer

After validating access and existence

> * Add URLs for the new view functions
> * Add "Approve" and "Reject" buttons to the release viewer page, but show only to output checkers and only when the request's state is `SUBMITTED`
>   
>   * If a user has already approved/rejected a file, only the button to reject/approve should be shown

Yeah, there's some nuance here. If a user updates the version of a file, we should reset all votes.

I *think* that should just work here, as with Dave's incoming immutable file changes, that will end up with a new RequestFile object, and thus will be reset back to 0 votes?

I'm not 100% sure that we want that, tbh. I think updating the file should invalidate all reviews on that file, and require you to review again, but we'd want to retain history.

And if I'm wrong about the implications of Dave's change, and its the *same* RequestFile instance, we need to have an `invalidated` field for FileApproval's, or similar, for went he file_id is updated.


> * Add `get_file_approvals` method to `LocalDBDataAccessLayer`
>   * this should take a file and return a list of dicts with the properties of file's approvals

See below - I suggest this should work differently

> * Add `FileApproval` dataclass to `business_logic.py` (cf `FileGroup` dataclass)
> * Add `get_file_approvals` method to `BusinessLogicLayer` which calls through to the DAL and returns a `FileApproval`

I would much prefer it if `business_logic.RequestFile` grew an `.approvals` attr, which is a list of the new `business_logic.FileApproval` objects for that fil. And that when get_release_request will build up the full tree of objects via its existing from_dict mechanism, and we can just iterate the one object, rather than integrating two data sources.

Basically, modelling the django db relationships in the business logic objects.

> * Support showing file approvals (who approved/rejected? when?) to the release viewer page

There are several potential places to display this info

1) in the nav tree, possibly by color or icon, to give an at-a-glance view, and easily navigate to files that still need approval.
2) in the file contents header (where other metdata and buttons live) 
3) in a directory listing (where I'd ho.pe we'll have multiselect approve/reject options in future). This could be deferred to later.

**Comment by inglesp** [comment]
Thanks @bloodearnest.  Will update the notes with most of your suggestions.  However, as it's not yet possible to update the version of a file in a request I don't think we should think about that here, as this is already quite a chunky piece of work.

**Comment by madwort** [comment]
> >         Add get_file_approvals method to BusinessLogicLayer which calls through to the DAL and returns a FileApproval
> 
> I would much prefer it if business_logic.RequestFile grew an .approvals attr, which is a list of the new business_logic.FileApproval objects for that fil. And that when get_release_request will build up the full tree of objects via its existing from_dict mechanism, and we can just iterate the one object, rather than integrating two data sources.

I agree & have done this!

**Comment by bloodearnest** [comment]
It would be good to indicate in the nav tree if a file is approved or rejected, possibly wiht some kind of green/red highlighting or icon.

That should make it easier for researchers to quickly navigate to problematic files, or for output-checkers to easily know what files still need review.

But this could be done in a later issue/PR, perhaps
